File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
import math

def solution(numbers):
    def generate_combinations(numbers):
        result = set()

        def backtrack(path, used):
            if path:
                num = int(''.join(path))
                result.add(num)

            for i in range(len(numbers)):
                if not used[i]:
                    used[i] = True
                    path.append(numbers[i])
                    backtrack(path, used)
                    path.pop()
                    used[i] = False

        used = [False] * len(numbers)
        backtrack([], used)

        return result

    numbers = list(numbers)
    result = generate_combinations(numbers)
    M = max(result)
    
    is_valid = [True] * (M + 1)
    is_valid[0] = is_valid[1] = False
    
    for i in range(2, math.isqrt(M) + 1):
        if is_valid[i]:
            for j in range(i * i, M + 1, i):
                is_valid[j] = False
    
    return sum(1 for i in result if is_valid[i])

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;
import java.lang.Math;

public class Solution {

    public int solution(String numbers) {
        Set<Integer> generated = generateCombinations(numbers.toCharArray());
        int max = Collections.max(generated);
        
        boolean[] isPrime = sieve(max);

        int count = 0;
        for (int n : generated) {
            if (isPrime[n]) count++;
        }
        return count;
    }

    // 모든 가능한 숫자 조합 생성
    private Set<Integer> generateCombinations(char[] numbers) {
        Set<Integer> result = new HashSet<>();
        boolean[] used = new boolean[numbers.length];
        backtrack(numbers, new StringBuilder(), used, result);
        return result;
    }

    private void backtrack(char[] numbers, StringBuilder path, boolean[] used, Set<Integer> result) {
        if (path.length() > 0) {
            int num = Integer.parseInt(path.toString());
            result.add(num);
        }

        for (int i = 0; i < numbers.length; i++) {
            if (!used[i]) {
                used[i] = true;
                path.append(numbers[i]);
                backtrack(numbers, path, used, result);
                path.deleteCharAt(path.length() - 1);
                used[i] = false;
            }
        }
    }

    // 에라토스테네스의 체
    private boolean[] sieve(int max) {
        boolean[] isPrime = new boolean[max + 1];
        Arrays.fill(isPrime, true);
        if (max >= 0) isPrime[0] = false;
        if (max >= 1) isPrime[1] = false;

        int sqrt = (int) Math.sqrt(max);
        for (int i = 2; i <= sqrt; i++) {
            if (isPrime[i]) {
                for (int j = i * i; j <= max; j += i) {
                    isPrime[j] = false;
                }
            }
        }
        return isPrime;
    }
}
